File: ios/Extopy/Repositories/Application/TokenRepository.swift
//
//  TokenRepository.swift
//  Extopy
//
//  Created by Nathan Fallet on 22/12/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import Keychain
import shared

class TokenRepository: ITokenRepository {
    
    private let keychain = Keychain()
    
    func getToken() -> String? {
        keychain.value(forKey: "token") as? String
    }

    func getRefreshToken() -> String? {
        keychain.value(forKey: "refreshToken") as? String
    }
    
    func getUserId() -> CoreUUID? {
        if let uuid = keychain.value(forKey: "userId") as? UUID {
            CoreUUID(nsUUID: uuid)
        } else {
            nil
        }
    }
    
    func setToken(token: String?) {
        let _ = if let token {
            keychain.save(token, forKey: "token")
        } else {
            keychain.remove(forKey: "token")
        }
    }

    func setRefreshToken(token: String?) {
        let _ = if let token {
            keychain.save(token, forKey: "refreshToken")
        } else {
            keychain.remove(forKey: "refreshToken")
        }
    }

    func setUserId(userId: CoreUUID?) {
        let _ = if let userId {
            keychain.save(userId.nsUUID, forKey: "userId")
        } else {
            keychain.remove(forKey: "userId")
        }
    }
    
}

File: android/src/main/kotlin/com/extopy/repositories/application/TokenRepository.kt
package com.extopy.repositories.application

import android.content.Context
import dev.kaccelero.models.UUID

class TokenRepository(
    context: Context,
) : ITokenRepository {

    private val sharedPreferences = context.getSharedPreferences("extopy", Context.MODE_PRIVATE)

    override fun getToken(): String? = sharedPreferences.getString("token", null)

    override fun setToken(token: String?) = sharedPreferences.edit()
        .putString("token", token)
        .apply()

    override fun getRefreshToken(): String? = sharedPreferences.getString("refreshToken", null)

    override fun setRefreshToken(token: String?) = sharedPreferences.edit()
        .putString("refreshToken", token)
        .apply()

    override fun getUserId(): UUID? =
        sharedPreferences.getString("userId", null)?.let(::UUID)

    override fun setUserId(userId: UUID?) = sharedPreferences.edit()
        .putString("userId", userId.toString())
        .apply()

}
